support multiple external C files having the same name

zerbina opened the pull request:
## Summary

Adding multiple C files sharing the same name to the build via the
`compile` pragma is now supported.

Closes https://github.com/nim-works/nimskull/issues/1291.

## Details

For the non-pattern `compile` pragma, the C file name is turned into a
unique object file name via `withPackageName`, mirroring what the C
backend does for the C files it generates.

To prevent the object files for the external C files from colliding
with the NimSkull module object files, they're placed under the
`external` sub-directory within the cache directory.

zerbina committed:
pragmas: ensure unique object output file names

Add the package name to the object file names to support compiling
external C files that share the same name.

zerbina committed:
tests: add a test

zerbina commented:
Hm, there's a problem I didn't account for: an external C file with the same name as a `.nim` using a `.nim.c` extension leads to a name collision of the object files.

One solution would be to simply dismiss this being a problem, a second one to mangle the name in a way that no collisions are possible. I think 'll give the second option a try.

**Edit:** There's an even simpler, yet more robust solution. Instead of putting the object files for external C files in the same directory as the normal object files, they can be placed in a sub-directory.

saem commented:
> **Edit:** There's an even simpler, yet more robust solution. Instead of putting the object files for external C files in the same directory as the normal object files, they can be placed in a sub-directory.

Having it separate makes a lot of sense, they're like second class compilation units. ;)

zerbina committed:
tests: add the missing C file

It should have been included in the earlier commit.

zerbina committed:
prevent collisions with object files for NimSkull modules

It's possible to add to the build external C files that use names for
which the generated object file would collide with that of a NimSkull
module's object file. To prevent this from happening, external C files
are now compiled to object files located in the `external` sub-
directory within the cache directory.

saem commented:
/merge
